crates/sui-framework/packages/sui-framework/sources/k_of_n_upgrade_policy.move

```diff
@@ -0,0 +1,395 @@
+            id: proposal_uid,
+            upgrade_cap: cap_id,
+            proposer,
+            digest,
+            current_voters: vec_set::empty(),
+        })
+    }
+
+    /// Vote in favor of an upgrade, aborts if the voter is not for the proposed
+    /// upgrade or if they voted already, or if the upgrade was already performed.
+    public fun vote(
+        proposal: &mut ProposedUpgrade, 
+        voter: &mut VotingCap,
+        ctx: &TxContext,
+    ) {
+        assert!(proposal.proposer != @0x0, EAlreadyIssued);
+        assert!(proposal.upgrade_cap == voter.upgrade_cap, EInvalidVoterForUpgrade);
+        let voter_id = object::id(voter);
+        assert!(
+            !vec_set::contains(&proposal.current_voters, &voter_id), 
+            EAlreadyVoted,
+        );
+        vec_set::insert(&mut proposal.current_voters, voter_id);
+        voter.votes_issued = voter.votes_issued + 1;
+
+        event::emit(UpgradeVoted {
+            proposal: object::id(proposal),
+            digest: proposal.digest,
+            voter: voter_id,
+            signer: tx_context::sender(ctx),
+        });
+    }
+
+    /// Issue an `UpgradeTicket` for the upgrade being voted on.  Aborts if 
+    /// there are not enough votes yet, or if the upgrade was already performed.
+    /// The signer of the transaction must be the same as the one proposing the
+    /// upgrade.
+    public fun authorize_upgrade(
+        cap: &mut KofNUpgradeCap,
+        proposal: &mut ProposedUpgrade, 
+        ctx: &TxContext,
+    ): UpgradeTicket {
+        assert!(proposal.upgrade_cap == object::id(cap), EInvalidProposalForUpgrade);
+        assert!(
+            vec_set::size(&proposal.current_voters) >= cap.required_votes, 
+            ENotEnoughVotes,
+        );
+        assert!(proposal.proposer != @0x0, EAlreadyIssued);
+
+        // assert the signer is the proposer and the upgrade has not happened yet
+        let signer = tx_context::sender(ctx);
+        assert!(proposal.proposer == signer, ESignerMismatch);
+        proposal.proposer = @0x0;
+
+        event::emit(UpgradePerformed {
+            upgrade_cap: proposal.upgrade_cap,
+            proposal: object::id(proposal),
+            digest: proposal.digest,
+            proposer: signer,
+        });
+
+        let policy = package::upgrade_policy(&cap.upgrade_cap);
+        package::authorize_upgrade(
+            &mut cap.upgrade_cap,
+            policy,
+            proposal.digest,
+        )
+    }
+
+    /// Finalize the upgrade to produce the given receipt.
+    public fun commit_upgrade(
+        cap: &mut KofNUpgradeCap, 
+        receipt: UpgradeReceipt,
+    ) {
+        package::commit_upgrade(&mut cap.upgrade_cap, receipt)
+    }
+
+    /// Discard an existing proposed upgrade.
+    /// The signer of the transaction must be the same address that proposed the
+    /// upgrade.
+    public fun discard_proposed_upgrade(proposed_upgrade: ProposedUpgrade, ctx: &TxContext) {
```

Comment:
(This API doesn't make sense unless we have shared object deletion)

Reply:
correct, and we could delete it until then or leave it here and it will start working when shared object deletion works.
I was keeping it here as it did not seen to harm anything but I am fine removing if there are strong feelings about not having it

Reply:
I don't think it's a big deal, but it's mildly confusing because there is a legitimate implementation of discarding a proposed upgrade (which sets the `proposer` to `@0x0`) that doesn't require shared object deletion, so it raises the question of why this is here and that implementation isn't.

We will have a similar confusion if we introduce new  APIs for shared object deletion after the fact as well, and at that point, I'm wondering whether it's worth publishing a brand new policy, and then upgrading the existing one to offer a migration function?

Reply:
It's true that one would pass the proposal by value in `authorize` and be done with it. That is the ideal scenario and the reason why shared object deletion is very important in my opinion. There is nothing like controlling operations on an object as destroying the object itself. It's a joy to have that semantics. 
But as you say we do not have that now and there is not much we can do.
This was here more as a way to recover some money for the object cost and to put an end to the object lifetime.
We could have a call to zero the proposer but I am not clear it gives you enough value, or at least, the proposer has to authorize the upgrade so not protecting much. Also I hope shared object deletion is coming soon enough to make this mostly a non issue and this API useful without an upgrade of any kind. Which is why it's here.
I am not sure about having an entirely new contract, it does not seem worth to me and it feels we could add a better API or leave with this `authorize` and `destroy` pattern which would be ok assuming delete shared object comes soon enough.
Well my thoughts at least ;)
but of course fine removing it and adding it later if we think it will be more of a problem than anything else